Added basic integration tests for remaining examples

strickvl opened the pull request:
## Describe changes
Added the remaining examples into our basic integration tests.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/features/integrations) table.
- [x] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

stefannica commented on examples/airflow_local/run_schedule.py:
```diff
@@ -0,0 +1,121 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+from datetime import datetime, timedelta
```
You could use one of two tricks to avoid duplicating the whole example code:

* move the common elements (steps, pipelines, utility functions and so on) to an `example.py` or `pipelines.py` file and import everything from it into `run.py` and `run_schedule.py`. This approach is used in the [mlflow_tracking example](https://github.com/zenml-io/zenml/tree/main/examples/mlflow_tracking) 
* use click to add a command line argument to `run.py` that controls which type of pipeline run is launched (scheduled vs. one-time)

strickvl replied:
Good point. Done.

stefannica commented on examples/dag_visualizer/run_visualize.py:
```diff
@@ -0,0 +1,131 @@
+#  permissions and limitations under the License.
+
+
+import pandas as pd
```
Same thing here: you should avoid duplicating the whole example code, if possible.

strickvl replied:
Done.

stefannica commented on tests/integration/test_examples.py:
```diff
@@ -123,6 +123,55 @@ class ExampleIntegrationTestConfiguration(NamedTuple):
+        ),
+    ),
+    ExampleIntegrationTestConfiguration(
+        name="functional_api",
```
Same here: how does the test framework know which file to run ?

stefannica commented on examples/lineage/run_visualize.py:
```diff
@@ -0,0 +1,159 @@
+#  permissions and limitations under the License.
+
+
+import pandas as pd
```
Same thing here: you should avoid duplicating the whole example code, if possible.

strickvl replied:
Done

stefannica commented on tests/integration/test_examples.py:
```diff
@@ -123,6 +123,55 @@ class ExampleIntegrationTestConfiguration(NamedTuple):
+        ),
+    ),
+    ExampleIntegrationTestConfiguration(
+        name="class_based_api",
```
'm curious, how does the test framework know which file to run for this example ? There is no run.py file, but a bunch of chapter_*.py files.

stefannica replied:
Ok, I think I got it, the function that selects which of the python files in an example is in zenml.cli.example:

```
    @property
    def executable_python_example(self) -> str:
        """Return the Python file for the example"""
        if self.has_single_python_file:
            return self.python_files_in_dir[0]
        elif self.run_dot_py_file:
            return self.run_dot_py_file
        elif self.has_any_python_file:
            logger.warning(
                "This example has multiple executable python files. "
                "The last one in alphanumerical order is taken."
            )
            return sorted(self.python_files_in_dir)[-1]
        else:
            raise RuntimeError(
                "No pipeline runner script found in example. "
                f"Files found: {self.python_files_in_dir}"
            )
```

In our case, the last one in alphanum order is executed.

strickvl replied:
The logic for this is in `src/zenml/cli/example.py`. We first check whether there is just one python file inside the example folder. If so, that's the one we run. Then if there are multiple files but one of them is called `run.py`, then that's the one we run. And finally if there are multiple files (but none of them are called `run.py`, when we just pick the last one in alphanumerical order to run.

stefannica reviewed (changes requested):
I left a couple of ideas to help with code duplication, but looks good otherwise

strickvl commented:
@stefannica addressed all your comments and refactored all the tests which have one version for testing and another for the purposes of showcasing the example.

stefannica reviewed (approved):
coverage +100% !

schustmi reviewed (approved):
Really nice!
